src/pykeen/models/resolve.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+
+"""Generation of ad-hoc model classes."""
+
+import logging
+from typing import Any, Mapping, Optional, Sequence, Tuple, Type, Union
+
+from .nbase import ERModel, EmbeddingSpecificationHint
+from ..nn import EmbeddingSpecification, RepresentationModule
+from ..nn.modules import Interaction, interaction_resolver
+from ..typing import Hint
+
+__all__ = [
+    'model_builder',
+    'model_from_interaction',
+]
+
+logger = logging.getLogger(__name__)
+
+
+def _normalize_entity_representations(
+    dimensions: Mapping[str, int],
+    interaction: Type[Interaction],
+    entity_representations: EmbeddingSpecificationHint,
+    relation_representations: EmbeddingSpecificationHint,
+) -> Tuple[Sequence, Sequence]:
+    if entity_representations is None:
+        # TODO: Does not work for interactions with separate tail_entity_shape (i.e., ConvE)
+        if interaction.tail_entity_shape is not None:
+            raise NotImplementedError
+        entity_representations = [
+            EmbeddingSpecification(shape=tuple(
+                dimensions[d]
+                for d in shape
+            ))
+            for shape in interaction.entity_shape
+        ]
+    if relation_representations is None:
+        relation_representations = [
+            EmbeddingSpecification(shape=tuple(
+                dimensions[d]
+                for d in shape
+            ))
+            for shape in interaction.relation_shape
+        ]
+    return entity_representations, relation_representations
+
+
+def model_instance_builder(
+    dimensions: Mapping[str, Any],
```

Comment:
@cthoyt this dictionary might be a bit unintuitive, since it uses the names (characters) from the interaction functions. Maybe we can support passing more natural names (or change the names of the interaction functions to more verbose ones)?

Reply:
Agreed, I can not really tell what this does at the moment